stdlib/LinearAlgebra/src/blas.jl

```diff
@@ -157,6 +160,58 @@ function check()
     end
 end
 
+"""
+    with_num_threads(f, num_threads::Integer)
+
+Run function `f()` with BLAS threads `num_threads` and then
+restore to previous threads setting.
+
+!!! compat "Julia 1.8"
+    `with_num_threads` requires at least Julia 1.8.
+
+# Example
+
+Depending on the number of available CPU cores, the result can be different:
+
+```julia
+julia> BLAS.get_num_threads()
+8
+
+julia> with_num_threads(4) do
+    BLAS.get_num_threads()
+    # or doing some basic BLAS computation
+end
+4
+
+julia> BLAS.get_num_threads()
+8
+```
+
+!!! warning
+    This function is not thread safe. If there are multiple
+    threads calling BLAS routines, then the threads they are
+    using will also be changed until this function finishes.
+
+!!! warning
+    This interface is experimental and subject to change or
+    removal without notice.
+
+See also [`set_num_threads`](@ref BLAS.set_num_threads) to permanently change
+number of BLAS threads.
+"""
+function with_num_threads(f, num_threads::Integer)
+    prev_num_threads = BLAS.get_num_threads()
+    BLAS.set_num_threads(num_threads)
+    local retval
+    try
+        retval = f()
+    catch err
```

Comment:
```suggestion
    catch 
```

and just use `rethrow()`?